### PR TITLE
feat: add precondition check middleware and handler headers

### LINKED ISSUE #78: Create precondition header middleware
To avoid doing queries we shouldn't, the precondition header check should be done in middleware in front of the db middlware, to verify whether the query has any chance of succeeding.

### PR DESCRIPTION
**Pull request by bbangert**
Most handler functions now have the appropriate header added. The
precondition middleware returns the request early if the modified
header and database differ.

Closes #78

### CODE REVIEW

File: src/web/middleware.rs [hunk 1]
```diff
@@ -131,6 +131,50 @@ impl Middleware<ServerState> for DbTransaction {
...
+        let collection = CollectionParam::from_request(&req, &())
+            .ok()
+            .map(|v| v.collection);
+        let bso = BsoParam::from_request(&req, &()).ok().map(|v| v.bso);
+        let fut = db
+            .extract_resource(user_id, collection, bso)
+            .and_then(move |resource_ts| {
+                let (status, modified) = match precondition {
+                    PreConditionHeader::IfModifiedSince(header_ts) if resource_ts <= header_ts => {
+                        (StatusCode::NOT_MODIFIED, header_ts)
```
Comment (pjenvey) [inline src/web/middleware.rs]: Unless I'm mistaken python's rendering resource_ts as the X-Last-Modified in these cases, not header_ts?

File: src/web/middleware.rs [hunk 2]
```diff
@@ -131,6 +131,50 @@ impl Middleware<ServerState> for DbTransaction {
...
+                    }
+                    PreConditionHeader::IfUnmodifiedSince(header_ts) if resource_ts > header_ts => {
+                        (StatusCode::PRECONDITION_FAILED, header_ts)
```
Comment (pjenvey) [inline src/web/middleware.rs]: same here

File: src/db/mysql/models.rs [hunk 3]
```diff
@@ -433,7 +432,7 @@ impl MysqlDb {
         Ok(results::GetBsos {
             bsos,
             more,
-            offset: next_offset,
+            offset: next_offset as i64,
```
Comment (pjenvey) [inline src/db/mysql/models.rs]: nit I forgot to revert this cast, it's no longer needed

File: src/web/middleware.rs (hunk 1 above)
Comment (bbangert) [reply src/web/middleware.rs]: This is purely the status code for the request, not the header.

Comment (pjenvey) [reply src/web/middleware.rs]: I meant the second item of these tuple set as X-Last-Modified below

File: src/db/mysql/models.rs (hunk 3 above)
Comment (bbangert) [reply src/db/mysql/models.rs]: Fixed.

File: src/web/middleware.rs (hunk 1 above)
Comment (bbangert) [reply src/web/middleware.rs]: Ah! Right.

**Pull request merged**